src/pairwise.jl

```diff
@@ -1,25 +1,53 @@
-function _pairwise!(::Val{:none}, f, dest::AbstractMatrix, x, y, symmetric::Bool)
-    @inbounds for (i, xi) in enumerate(x), (j, yj) in enumerate(y)
-        symmetric && i > j && continue
-
-        # For performance, diagonal is special-cased
-        if f === cor && eltype(dest) !== Union{} && i == j && xi === yj
-            # TODO: float() will not be needed after JuliaLang/Statistics.jl#61
-            dest[i, j] = float(cor(xi))
-        else
-            dest[i, j] = f(xi, yj)
-        end
+function _pairwise!(::Val{:none}, f, dest::AbstractMatrix{V}, x, y,
+    symmetric::Bool) where {V}
+
+    nr, nc = size(dest)
+
+    # Swap x and y for more efficient threaded loop.
+    if nr < nc
+        dest′ = reshape(dest, size(dest, 2), size(dest, 1))
+        _pairwise!(Val(:none), f, dest′, y, x, symmetric)
+        dest .= transpose(dest′)
+        return dest
     end
-    if symmetric
-        m, n = size(dest)
-        @inbounds for j in 1:n, i in (j+1):m
-            dest[i, j] = dest[j, i]
+
+    #cor and friends are special-cased.
+    iscor = (f in (corkendall, corspearman, cor))
+    (iscor || f == cov) && (symmetric = x === y)
+    #cov(x) is faster than cov(x, x)
+    (f == cov) && (f = ((x, y) -> x === y ? cov(x) : cov(x, y)))
+
+    Threads.@threads for subset in equal_sum_subsets(nr, Threads.nthreads())
```

Comment:
Is it clear that this way to split the iteration is (somewhat) optimal? For instance, I would assume that it would be beneficial if each iteration/task would operate on contiguous memory regions which doesn't seem to be the case here.

I wonder if it is even beneficial to pre-partition the iteration space here in smaller subsets or whether one would achieve the same performance gains by only performing a threaded loop over the upper-triangular indices.

Reply:
The situation I have in mind is when `f` is expensive to calculate (as is the case for `corkendall` and `corspearman`) and so the important thing is to ensure that each thread makes the same number of calls to `f`, so as to have all threads finish their work near simultaneously. I was able to see the benefit of `equal_sum_subsets`  when calculating a 5,000 x 5,000 kendall correlation matrices simply by observing the loading of cores in Windows task manager. My prior approach had a nasty tail of only one or two cores doing any work. `equal_sum_subsets` solved that with about a 25% speedup (if I recall correctly).

I also experimented with with using [ChunkSplitters](https://github.com/JuliaFolds2/ChunkSplitters.jl) with `split = : scatter` but that approach does not do quite as good a job of equal division between threads of calls to `f`. 

Surely considerations of memory layout would only come into play if `f` is a _really_ fast function? `LinearAlgebra.dot` might be a good example and I was very happy with the performance in that case.

In conclusion I doubt that there exists a significantly faster approach.